sw/device/lib/testing/rand_testutils.c

```diff
@@ -7,23 +7,75 @@
 #include <stdbool.h>
 #include <stdint.h>
 
+#include "sw/device/lib/arch/device.h"
+#include "sw/device/lib/dif/dif_rv_core_ibex.h"
+#include "sw/device/lib/runtime/log.h"
+#include "sw/device/lib/testing/rv_core_ibex_testutils.h"
 #include "sw/device/lib/testing/test_framework/check.h"
 
+#include "hw/top_earlgrey/sw/autogen/top_earlgrey.h"
+
 /**
  * The polynomial co-efficients used in the 32-bit LFSR implementation.
  *
  * This implementation matches the RTL design at `hw/ip/prim/rtl/prim_lfsr.sv`.
  */
-const uint32_t kLfsrCoefficients = 0x80000057;
+static const uint32_t kLfsrPolynomialCoefficients = 0x80000057;
 
-uint32_t rand_testutils_gen32(void) {
-  static uint32_t lfsr = 1;
-  bool lsb = lfsr & 0x1u;
-  lfsr >>= 1;
+/**
+ * The default timeout in usecs for fetching data from the entropy source.
+ */
+static const uint32_t kEntropyFetchTimeoutMicros = 1000;
+
+rand_testutils_rng_t rand_testutils_init(dif_rv_core_ibex_t *rv_core_ibex) {
+  CHECK(rv_core_ibex != NULL);
+  // For the simulation platforms (DV and Verilator), the LFSR reseed frequency
+  // is arbitrarily set to 255, i.e. after every 255 PRNG cycles, the LFSR is
+  // freshly reseeded from the on-device entropy source. The test may choose to
+  // update this value as needed.
+  rand_testutils_rng_t ctx = (rand_testutils_rng_t){
+      .rv_core_ibex = rv_core_ibex,
+      .entropy_fetch_timeout_usec = kEntropyFetchTimeoutMicros,
+      .lfsr = 0xdeadbeef,  // Arbitrary.
+      .polynomial_coefficients = kLfsrPolynomialCoefficients,
+      .reseed_frequency = 256,
+      .op_counter = UINT32_MAX};
+  // For non-runtime-sensitive simulation platforms (FPGA / debug board),
+  // always fetch random data from the entropy source.
+  if (kDeviceType != kDeviceSimDV && kDeviceType != kDeviceSimVerilator) {
+    ctx.reseed_frequency = 0;
```

Comment:
so if i understand correctly, if on silicon / fpga we will always invoke reseed lfsr, which will get us real entropy from the ibex random registers, is that right?

Reply:
Yes that is correct.